pkg/pool-weighted/contracts/InvariantGrowthProtocolFees.sol

```diff
@@ -25,6 +27,57 @@ abstract contract InvariantGrowthProtocolFees is BaseWeightedPool {
     // from performing any computation or accounting associated with protocol fees during swaps.
     // This mechanism requires keeping track of the invariant after the last join or exit.
     uint256 private _lastPostJoinExitInvariant;
+    uint256 private _lastRateGrowthProduct;
+
+    // Rate providers are used only for computing yield fees; they do not inform swap/join/exit.
+    IRateProvider internal immutable _rateProvider0;
+    IRateProvider internal immutable _rateProvider1;
+    IRateProvider internal immutable _rateProvider2;
+    IRateProvider internal immutable _rateProvider3;
+    IRateProvider internal immutable _rateProvider4;
+    IRateProvider internal immutable _rateProvider5;
+    IRateProvider internal immutable _rateProvider6;
+    IRateProvider internal immutable _rateProvider7;
+    IRateProvider internal immutable _rateProvider8;
+    IRateProvider internal immutable _rateProvider9;
+    IRateProvider internal immutable _rateProvider10;
+    IRateProvider internal immutable _rateProvider11;
+    IRateProvider internal immutable _rateProvider12;
+    IRateProvider internal immutable _rateProvider13;
+    IRateProvider internal immutable _rateProvider14;
+    IRateProvider internal immutable _rateProvider15;
+    IRateProvider internal immutable _rateProvider16;
+    IRateProvider internal immutable _rateProvider17;
+    IRateProvider internal immutable _rateProvider18;
+    IRateProvider internal immutable _rateProvider19;
+
+    constructor(IRateProvider[] memory rateProviders) {
+        uint256 numTokens = rateProviders.length;
```

Comment:
This is insufficient: we much check that `rateProviders` has the same lenght as the `tokens` array. Perhaps we could pass a `totalTokens` value to this constructor, and call it with e.g. `tokens.length`.

Reply:
(note that we cannot access `_getTotalTokens()` during construction)

Reply:
Yeah, I included that in the `WeightedPool` constructor because of the access issues

Reply:
Would you prefer passing the length into this constructor and performing the check here?

Reply:
Yes, I think that'd be less error prone. You can't forget to pass a constructor argument.